syncer/syncer.go

```diff
@@ -0,0 +1,272 @@
+}
+
+func (s *syncer) GetSyncProgression() *progress.Progression {
+	return s.syncProgression.GetProgression()
+}
+
+// HasSyncPeer returns whether syncer has the peer to syncs blocks
+// return false if syncer has no peer whose latest block height doesn't exceed local height
+func (s *syncer) HasSyncPeer() bool {
+	if s.peerMap == nil {
+		return false
+	}
+
+	bestPeer := s.peerMap.BestPeer(nil)
+	header := s.blockchain.Header()
+
+	return bestPeer != nil && bestPeer.Number > header.Number
+}
+
+func (s *syncer) BulkSync(ctx context.Context, newBlockCallback func(*types.Block) bool) error {
+	localLatest := uint64(0)
+	updateLocalLatest := func() {
+		if header := s.blockchain.Header(); header != nil {
+			localLatest = header.Number
+		}
+	}
+
+	updateLocalLatest()
+
+	// Create a blockchain subscription for the sync progression and start tracking
+	s.syncProgression.StartProgression(localLatest+1, s.blockchain.SubscribeEvents())
+
+	// Stop monitoring the sync progression upon exit
+	defer s.syncProgression.StopProgression()
+
+	skipList := make(map[peer.ID]bool)
+
+	for {
+		bestPeer := s.peerMap.BestPeer(skipList)
+		if bestPeer == nil || bestPeer.Number <= localLatest {
+			break
+		}
+
+		// Set the target height
+		s.syncProgression.UpdateHighestProgression(bestPeer.Number)
+
+		lastNumber, _, err := s.bulkSyncWithPeer(bestPeer.ID, newBlockCallback)
+		if err != nil {
+			s.logger.Warn("failed to complete bulk sync with peer, try to next one", "peer ID", bestPeer.ID)
+		}
+
+		// if node could sync with the peer fully, then exit loop
+		if err == nil && lastNumber >= bestPeer.Number {
+			break
+		}
+
+		updateLocalLatest()
+
+		skipList[bestPeer.ID] = true
+	}
+
+	return nil
+}
+
+func (s *syncer) WatchSync(ctx context.Context, callback func(*types.Block) bool) error {
+	localLatest := s.blockchain.Header().Number
+	skipList := make(map[peer.ID]bool)
+
+	// Loop until context is canceled
+	for {
+		//Wait for a new event to arrive
+		<-s.newStatusCh
+
+		// fetch local latest block
+		if header := s.blockchain.Header(); header != nil {
+			localLatest = header.Number
+		}
+
+		// pick one best peer
+		bestPeer := s.peerMap.BestPeer(skipList)
+		if bestPeer == nil {
+			// Empty skipList map if there are no best peers
+			skipList = make(map[peer.ID]bool)
+
+			continue
+		}
+
+		// if the bestPeer does not have a new block continue
+		if bestPeer.Number <= localLatest {
+			continue
+		}
+
+		// fetch block from the peer
+		lastNumber, shouldTerminate, err := s.bulkSyncWithPeer(bestPeer.ID, callback)
+		if err != nil {
+			s.logger.Warn("failed to complete bulk sync with peer, try to next one", "peer ID", bestPeer.ID)
+		}
+
+		if err != nil || lastNumber < bestPeer.Number {
+			skipList[bestPeer.ID] = true
+
+			// continue to next peer
+			continue
+		}
+
+		if shouldTerminate {
+			break
+		}
+	}
+
+	return nil
+}
+
+func (s *syncer) bulkSyncWithPeer(peerID peer.ID, newBlockCallback func(*types.Block) bool) (uint64, bool, error) {
```

Comment:
Can we not have this method return 50 different things? 🙃

Reply:
I don't understand. Please explain correct info

Reply:
I don't think it's wrong. But we need to consider defining struct when returning 4+ values

Reply:
cc: @Aleksao998

Reply:
I also prefer having max 2 attributes as the return value, but here was either this or duplicating the same method. Another solution was to define struct as Yoshiki said but do not think it is necessary.
@zivkovicmilos Can you please suggest how you think we should refactor this

Reply:
I think the cleanest way, if we can't break up the method itself, is to introduce a return struct.

@dbrajovic what do you think?

Reply:
I don't know why to returning struct instead of tuple here. It gives more complexity. There are some methods that return 3 values in other place... it's fuzzy

Reply:
The problem is that we didn't have threshold. But in my opinion, returning 3 values is acceptable.

Reply:
@zivkovicmilos I don't see the way to break this method and not to duplicate code, also i agree with @Kourin1996 that in this case 3 values are acceptable. If you are strongly not against it i would leave it as it is :)